chatmaild/src/chatmaild/expire.py

```diff
@@ -0,0 +1,171 @@
+"""
+Expire old messages and addresses.
+
+"""
+
+import os
+import shutil
+import sys
+import time
+from argparse import ArgumentParser
+from collections import namedtuple
+from datetime import datetime
+from stat import S_ISREG
+
+from chatmaild.config import read_config
+
+FileEntry = namedtuple("FileEntry", ("relpath", "mtime", "size"))
+
+
+def iter_mailboxes(basedir, maxnum):
+    for name in os.listdir(basedir)[:maxnum]:
+        if "@" in name:
+            yield MailboxStat(basedir + "/" + name)
+
+
+class MailboxStat:
+    last_login = None
+
+    def __init__(self, basedir):
+        self.basedir = str(basedir)
+        # all detected messages in cur/new/tmp folders
+        self.messages = []
+
+        # all detected files in mailbox top dir
+        self.extrafiles = []
+
+        # scan all relevant files (without recursion)
+        old_cwd = os.getcwd()
+        os.chdir(self.basedir)
+        for name in os.listdir("."):
+            if name in ("cur", "new", "tmp"):
+                for msg_name in os.listdir(name):
+                    relpath = name + "/" + msg_name
+                    st = os.stat(relpath)
+                    self.messages.append(FileEntry(relpath, st.st_mtime, st.st_size))
+            else:
+                st = os.stat(name)
+                if S_ISREG(st.st_mode):
+                    self.extrafiles.append(FileEntry(name, st.st_mtime, st.st_size))
+                    if name == "password":
+                        self.last_login = st.st_mtime
+        self.extrafiles.sort(key=lambda x: -x.size)
+        os.chdir(old_cwd)
+
+
+def print_info(msg):
+    print(msg, file=sys.stderr)
+
+
+class Expiry:
+    def __init__(self, config, dry, now, verbose):
+        self.config = config
+        self.dry = dry
+        self.now = now
+        self.verbose = verbose
+        self.del_mboxes = 0
+        self.all_mboxes = 0
+        self.del_files = 0
+        self.all_files = 0
+        self.start = time.time()
+
+    def remove_mailbox(self, mboxdir):
+        if self.verbose:
+            print_info(f"removing {mboxdir}")
+        if not self.dry:
+            shutil.rmtree(mboxdir)
+        self.del_mboxes += 1
+
+    def remove_file(self, path):
+        if self.verbose:
+            print_info(f"removing {path}")
+        if not self.dry:
+            try:
+                os.unlink(path)
+            except FileNotFoundError:
+                print_info(f"file not found/vanished {path}")
+        self.del_files += 1
```

Comment:
Do we have tests that this only unlinks files for one user, in case others have a valid reason to keep the file?

Right now we use hardlinks for deduplicating messages which are in more than one mailbox. the mtime will be the same for all of them, so for active mailboxes they should expire at the same time. But if one mboxdir is expiring and removed, this shouldn't delete the messages in all the other mailboxes.